### ISSUE TITLE
object-curly-newline for named imports not working as expected

### ISSUE THREAD

**Issue by antoniogiordano** [eslint/eslint#12018]
**Tell us about your environment**

* **ESLint Version:**
6.1.0
* **Node Version:**
12.2
* **npm Version:**
6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default through webstorm

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "parser": "babel-eslint",
  "extends": [
    "airbnb"
  ],
  "env": {
    "browser": true
  },
  "rules": {
    "arrow-parens": "off",
    "comma-dangle": ["error", "never"],
    "semi": ["error", "never"],
    "no-confusing-arrow": "off",
    "no-unused-expressions": "off",
    "react/jsx-filename-extension": [1, { "extensions": [".js", ".jsx"] }],
    "react/jsx-one-expression-per-line": "off",
    "import/prefer-default-export": "off",
    "no-param-reassign": "off",
    "no-unused-vars": "off",
    "import/extensions": "off",
    "react/no-unused-prop-types": "off",
    "object-curly-newline": ["error", {
      "ObjectExpression": {"minProperties": 1},
      "ObjectPattern": {"minProperties": 1},
      "ImportDeclaration": {
        "multiline": true,
        "minProperties": 1,
        "consistent": true
      },
      "ExportDeclaration": "always"
    }],
    "object-property-newline": ["error", {
      "allowAllPropertiesOnSameLine": false
    }]
  },
  "settings": {
    "import/resolver": {
      "webpack": {
        "config": "configs/webpack.config.dev.js"
      }
    }
  },
  "settings": {
    "import/resolver": {
      "webpack": {
        "config": "configs/webpack.config.dev.js"
      }
    }
  },
  "plugins": [
    "react"
  ]
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```bash
automatically done by webstorm
```

**What did you expect to happen?**
I expect this code
```js
import {
  Button, Divider,
  Grid, LinearProgress, Link, MuiThemeProvider, Paper, Typography, withStyles
} from '@material-ui/core'
```

**What actually happened? Please include the actual, raw output from ESLint.**
to be like this
```js
import {
  Button, 
  Divider,
  Grid,
  LinearProgress, 
  Link, 
  MuiThemeProvider, 
  Paper, 
  Typography, 
  withStyles
} from '@material-ui/core'
```

**Are you willing to submit a pull request to fix this bug?**
No

**Comment by kaicataldo** [comment]
It's not clear to me if this is reporting an issue with auto fixing or a lint error. Do you mind clarifying, and if it is a lint error, share the output of that error?

Additionally, to clarify whether it's an issue with ESLint itself or the Webstorm ESLint integration, can you run ESLint from the command line and confirm that it yields the same results?

**Comment by antoniogiordano** [comment]
Yes autofixing is enabled, even launching the command from the console there is no fix in the import statement.
The error is referred to the object-curly-newline rule, because it doesn't put all the import vars on new lines like I would expect

**Comment by mastacheata** [comment]
I can probably expand a little on this, but only using the 5.16.0 version of ESLint and using babel-eslint as the parser. (limited to 5.16.0 as we're using create-react-app on the project and that has a hard requirement for the 5.x series of ESLint)

This is the simplified rule from my .eslintrc.js:
```
    'object-curly-newline': ['error', {
        "ImportDeclaration": { "minProperties": 3, "consistent": false, "multiline": true },
    }],
```

According to the example on the ESLint documentation ( https://eslint.org/docs/rules/object-curly-newline#multiline) the following line should raise an error with ESLint as it has a)  >= 3 named imports and b) all of them on the very same line, which should be forbidden by the multiline: true setting:
```
import {
  ButtonControl, DatePickerControl, SelectControl, TextboxControl,
} from '../../../../shared/Form/elements';
```

I would expect this to be reformatted to have each import on it's own line:
```
import {
  ButtonControl,
  DatePickerControl,
  SelectControl,
  TextboxControl,
} from '../../../../shared/Form/elements';
```

Calling eslint from the commandline to analyze just one file like this:
` .\node_modules\.bin\eslint --ext=jsx ./project/index.js`

I also tried changing the object-property-newline rule to disable the `allowAllPropertiesOnSameLine` option, but that doesn't seem to have any effect at all.

If I change the conflicting import to be just a single line it will complain, but only move the named imports to the format specified above on a single line with the parenthesis on separate lines.

**Comment by phakphumi** [comment]
Same issue, even @mastacheata can't fix mine.

The lint can't enforce the rule to separate line from destructuring import. Like
```
import {
  Col, Row
} from 'antd';
```

Above code pass lint easily no warning no fixing.
My rules:
```
    "object-curly-newline": ["error", { "ImportDeclaration": { "multiline": true, "minProperties": 2 }}],
    "object-property-newline": "error",
```

**Comment by mneumark** [comment]
Same here. On my project, we want to be able to see github blame for each named import. To do that we need developers to put each named import on a separate line. So the ability to enforce one named import per line would be amazing.

**Comment by mi-mazouz** [comment]
+1

**Comment by brkn** [comment]
Is there any plan on fixing this?

**Comment by mastacheata** [comment]
Obviously not, The issue is closed, so noone will have it on their radar for things to fix.
Unless anyone affected by this problem has enough knowledge about ESLint internals to at least draft a PR, this is unlikely to ever be fixed.

**Comment by mdjermanovic** [comment]
Reopening to clarify, as this doesn't look like a bug.

`object-curly-newline` rule enforces line breaks after `{` and before `}`, i.e. between `{` and the first property and between the last property and `}`. 

`object-curly-newline` does work with named imports, and from the examples above it seems that it works as intended (please feel free to post again an example where it doesn't work well if I'm missing something). This rule does not enforce line breaks between properties inside `{}`.

The rule to enforce line breaks between properties is `object-property-newline`, but it works only with object literals. There is an ongoing discussion in #9259 on whether this rule should be enhanced to support destructuring patterns/imports/exports as well.

**Comment by mastacheata** [comment]
But isn't the Multiline option of that rule exactly for this situation?
Also, IIRC (not at a computer to double check right now) if you have more items on one line, ESLint complains and offers to fix it, but doesn't fix it...
At least to me,that is in fact a bug.

**Comment by mdjermanovic** [comment]
> But isn't the Multiline option of that rule exactly for this situation?

`multiline` option in `object-curly-newline` requires or disallows linebreaks after `{` and before `}` depending on whether the content inside is multiline or not. It doesn't enforce line breaks between properties/names inside `{}`. Line breaks between (or inside) properties/names are 'input' to determine what should be enforced on braces.

> Also, IIRC (not at a computer to double check right now) if you have more items on one line, ESLint complains and offers to fix it, but doesn't fix it...
> At least to me,that is in fact a bug.

Could you please post an example where it seems that the rule doesn't work as expected? That would be helpful to find a potential bug or clarify what's the responsibility of this rule.

**Comment by mastacheata** [comment]
I'll gladly provide an example on Monday afternoon when I'm in the office again.

**Comment by lukenofurther** [comment]
@mdjermanovic if this isn't a bug then [the documentation for the rule](https://eslint.org/docs/rules/object-curly-newline#importdeclaration-and-exportdeclaration) isn't accurate. According to the documentation:

``` js
// With this rule config in .eslintrc.js
"object-curly-newline": ["error", {
  "ImportDeclaration": "always",
}]

// operating on this subject statement
import { foo, bar } from "lodash"

// the rule's documentation implies that eslint --fix will change the subject to
import {
  foo,
  bar
} from "lodash"

// However, the actual result of eslint --fix is
import {
    foo, bar,
} from "lodash"
```
This is different from the implied "correct code" example in the rule's documentation.

The result of the ES Lint fix is still breaking [AirBnB styleguide rule 10.8](https://github.com/airbnb/javascript#modules--multiline-imports-over-newlines) so I'm surprised you won't fix it.

If you don't fix it then please at least update the documentation to reflect what --fix really does, and/or prevent this rule from being fixed for import statements at all.

**Comment by mdjermanovic** [comment]
@lukenofurther thanks for the suggestions!

Correct examples do not intend to represent the result of `eslint --fix` applied to incorrect examples.

`"object-curly-newline": ["error", { "ImportDeclaration": "always" }]` expects a line break after `{`, and a line break before `}`.

Any code that satisfies these conditions is correct for this rule:

```js
// correct example
import {
  foo,
  bar
} from "lodash"

// also correct example
import {
    foo, bar
} from "lodash"

// incorrect example
import { foo, bar } from "lodash"

// incorrect example
import { foo, 
bar } from "lodash"

// incorrect example
import { foo, bar
} from "lodash"

// incorrect example
import { 
foo, 
bar } from "lodash"

```


The documentation is correct in this case, but it would be nice to have both examples (with and without line breaks between `foo` and `bar`). PR is welcome!

`--fix` works as intended, because it makes a minimal change in the code to make it valid for this rule.

As for the line breaks between `foo` and `bar`, we don't have a rule to enforce/disallow these at the moment. Feel free to open a new rule proposal!
